HeterogeneousCore/CUDAUtilities/interface/GPUSimpleVector.h

```diff
@@ -0,0 +1,94 @@
+//  author: Felice Pantaleo, CERN, 2018
+#ifndef GPU_SIMPLEVECTOR_H_
+#define GPU_SIMPLEVECTOR_H_
+
+#include <cuda.h>
+#include <type_traits>
+#include <utility>
+
+namespace GPU {
+template <class T> struct SimpleVector {
+  // Constructors
+  __host__ __device__ SimpleVector(int capacity, T *data) // ownership of m_data stays within the caller
+      : m_size(0), m_data(data), m_capacity(capacity) {
+    static_assert(std::is_trivially_destructible<T>::value);
+  }
+
+  __host__ __device__ SimpleVector() : SimpleVector(0) {}
+
+  __inline__ __host__ __device__ int push_back_unsafe(const T &element) {
+    auto previousSize = m_size;
+    m_size++;
+    if (previousSize < m_capacity) {
+      m_data[previousSize] = element;
+      return previousSize;
+    } else {
+      --m_size;
+      return -1;
+    }
+  }
+
+  template <class... Ts> __host__ __device__ int emplace_back_unsafe(Ts &&... args) {
+    auto previousSize = m_size;
+    m_size++;
+    if (previousSize < m_capacity) {
+      (new (&m_data[previousSize]) T(std::forward<Ts>(args)...));
+      return previousSize;
+    } else {
+      --m_size;
+      return -1;
+    }
+  }
+
+  __inline__ __host__ __device__ T & back() const {
+
+    if (m_size > 0) {
+      return m_data[m_size - 1];
+    } else
+      return T(); //undefined behaviour
```

Comment:
Could this be simply `return m_data[m_size-1];`? For `m_size <= 0` the behaviour will be undefined anyway (leading to segfault if we're lucky), so this would allow removing one `if`.

Reply:
OK by me.

In fact it make little sense to return a temporary `T()` by reference.

And by the way, we should have both
```c++
T const & back() const;
T & back();
```